File: data/Demonstration/PD_C/ConnectFour.py
class ConnectFour:
    def __init__(self, numRows, numCols):
        self.numRows = numRows
        self.numCols = numCols
        self.board = [['-' for _ in range(numCols)] for _ in range(numRows)]
        self.currentPlayer = 'X'

    def initializeBoard(self):
        for i in range(self.numRows):
            for j in range(self.numCols):
                self.board[i][j] = '-'

    def printBoard(self):
        for i in range(self.numRows - 1, -1, -1):
            for j in range(self.numCols):
                print(self.board[i][j], end=' ')
            print()
        print()

    def isValidMove(self, col):
        if col < 0 or col >= self.numCols:
            return False
        if self.board[self.numRows - 1][col] != '-':
            return False
        return True

    def makeMove(self, col):
        for i in range(self.numRows):
            if self.board[i][col] == '-':
                self.board[i][col] = self.currentPlayer
                break

    def switchPlayer(self):
        if self.currentPlayer == 'X':
            self.currentPlayer = 'O'
        else:
            self.currentPlayer = 'X'

    def checkForWin(self):
        # check for a horizontal win
        for row in range(self.numRows):
            for col in range(self.numCols - 3):
                if self.board[row][col] != '-' and self.board[row][col] == self.board[row][col + 1] == self.board[row][col + 2] == self.board[row][col + 3]:
                    return True

        # check for a vertical win
        for row in range(self.numRows - 3):
            for col in range(self.numCols):
                if self.board[row][col] != '-' and self.board[row][col] == self.board[row + 1][col] == self.board[row + 2][col] == self.board[row + 3][col]:
                    return True

        # check for a diagonal win (top-left to bottom-right)
        for row in range(self.numRows - 3):
            for col in range(self.numCols - 3):
                if self.board[row][col] != '-' and self.board[row][col] == self.board[row + 1][col + 1] == self.board[row + 2][col + 2] == self.board[row + 3][col + 3]:
                    return True

        # check for a diagonal win (bottom-left to top-right)
        for row in range(3, self.numRows):
            for col in range(self.numCols - 3):
                if self.board[row][col] != '-' and self.board[row][col] == self.board[row - 1][col + 1] == self.board[row - 2][col + 2] == self.board[row - 3][col + 3]:
                    return True

        # no win found
        return False


    def checkForDraw(self):
        # check if the board is full
        for row in range(self.numRows):
            for col in range(self.numCols):
                if self.board[row][col] == '-':
                    return False

        # board is full, game is a draw
        return True


    def play(self):
        input_func = input
        gameOver = False
        while not gameOver:
            self.printBoard()
            col = input_func("Player {}, please enter your move (0-{}): ".format(self.currentPlayer, self.numCols - 1))
            try:
                col = int(col)
                if not self.isValidMove(col):
                    print("Invalid move. Please try again.")
                    continue
                self.makeMove(col)
                if self.checkForWin():
                    print("Player", self.currentPlayer, "wins!")
                    gameOver = True
                elif self.checkForDraw():
                    print("Game over. It's a draw.")
                    gameOver = True
                else:
                    self.switchPlayer()
            except ValueError:
                print("Invalid input")

        self.printBoard()


if __name__ == '__main__':
    game = ConnectFour(6, 7)
    game.play()

File: data/Validation/Java Blackjack/JD_C/ConnectFour.java
import java.util.Scanner;

public class ConnectFour {
	private char[][] board; // 2D array to represent the game board
	private int numRows; // number of rows on the game board
	private int numCols; // number of columns on the game board
	private char currentPlayer; // current player (either 'X' or 'O')

	public ConnectFour(int numRows, int numCols) {
		this.numRows = numRows;
		this.numCols = numCols;
		board = new char[numRows][numCols];
		currentPlayer = 'X';
		initializeBoard();
	}

	private void initializeBoard() {
		// set every cell on the board to '-'
		for (int i = 0; i < numRows; i++) {
			for (int j = 0; j < numCols; j++) {
				board[i][j] = '-';
			}
		}
	}

	private void printBoard() {
		// print out the game board
		for (int i = numRows - 1; i >= 0; i--) {
			for (int j = 0; j < numCols; j++) {
				System.out.print(board[i][j] + " ");
			}
			System.out.println();
		}
		System.out.println();
	}

	private boolean isValidMove(int col) {
		// check if the given column is a valid move
		if (col < 0 || col >= numCols) {
			return false;
		}
		if (board[numRows - 1][col] != '-') {
			return false;
		}
		return true;
	}

	private void makeMove(int col) {
		// make a move in the given column
		for (int i = 0; i < numRows; i++) {
			if (board[i][col] == '-') {
				board[i][col] = currentPlayer;
				break;
			}
		}
	}

	private boolean checkForWin() {
		// check for a win on the game board
		// check rows
		for (int i = 0; i < numRows; i++) {
			for (int j = 0; j <= numCols - 4; j++) {
				if (board[i][j] == currentPlayer && board[i][j + 1] == currentPlayer && board[i][j + 2] == currentPlayer
						&& board[i][j + 3] == currentPlayer) {
					return true;
				}
			}
		}
		// check columns
		for (int j = 0; j < numCols; j++) {
			for (int i = 0; i <= numRows - 4; i++) {
				if (board[i][j] == currentPlayer && board[i + 1][j] == currentPlayer && board[i + 2][j] == currentPlayer
						&& board[i + 3][j] == currentPlayer) {
					return true;
				}
			}
		}
		// check diagonals
		for (int i = 0; i <= numRows - 4; i++) {
			for (int j = 0; j <= numCols - 4; j++) {
				if (board[i][j] == currentPlayer && board[i + 1][j + 1] == currentPlayer
						&& board[i + 2][j + 2] == currentPlayer && board[i + 3][j + 3] == currentPlayer) {
					return true;
				}
			}
		}
		for (int i = 0; i <= numRows - 4; i++) {
			for (int j = numCols - 1; j >= 3; j--) {
				if (board[i][j] == currentPlayer && board[i + 1][j - 1] == currentPlayer
						&& board[i + 2][j - 2] == currentPlayer && board[i + 3][j - 3] == currentPlayer) {
					return true;
				}
			}
		}
		return false;
	}

	private boolean checkForDraw() {
		// check if the game is a draw
		for (int i = 0; i < numRows; i++) {
			for (int j = 0; j < numCols; j++) {
				if (board[i][j] == '-') {
					return false;
				}
			}
		}
		return true;
	}

	private void switchPlayer() {
		// switch the current player
		if (currentPlayer == 'X') {
			currentPlayer = 'O';
		} else {
			currentPlayer = 'X';
		}
	}

	public void play() {
		Scanner scanner = new Scanner(System.in);
		boolean gameOver = false;
		while (!gameOver) {
			// print out the game board
			printBoard();
			// get the current player's move
			System.out.println("Player " + currentPlayer + ", please enter your move (0-" + (numCols - 1) + "): ");
			try {
				int col = scanner.nextInt();
				// validate the move
				if (!isValidMove(col)) {
					System.out.println("Invalid move. Please try again.");
					continue;
				}
				// make the move
				makeMove(col);
				// check for a win
				if (checkForWin()) {
					System.out.println("Player " + currentPlayer + " wins!");
					gameOver = true;
				} else if (checkForDraw()) {
					System.out.println("Game over. It's a draw.");
					gameOver = true;
				} else {
					// switch the player
					switchPlayer();
				}
			} catch (Exception e) {
				System.out.println("Invalid input");
				scanner.nextLine();
			}
		}
		// print out the final game board
		printBoard();
		scanner.close();
	}

	public static void main(String[] args) {
		ConnectFour game = new ConnectFour(6, 7);
		game.play();
	}
}
